docs/SPIRV-Cookbook.rst

```diff
@@ -0,0 +1,976 @@
+images must be done directly on the global resources. They cannot be
+copied or have their address returned from functions. However, in HLSL,
+it is possible to copy a global resource or to pass it by reference to a
+function. Since this can be arbitrarily complex, DXC can generate valid
+SPIR-V only if the compiler is able to remove all of these copies.
+
+The transformations that are used to remove the copies will be the same
+for both structured buffers and images, so we have chosen to focus on
+structured buffer. The process of transforming the code in this way is
+called *legalization.*
+
+Support evolves over time as the optimizations in SPIRV-Tools are
+improved. At GDC 2018, Greg Fischer from LunarG
+`presented <http://schedule.gdconf.com/session/hlsl-in-vulkan-there-and-back-again-presented-by-khronos-group/856616>`__
+earlier results in this space. The DXC, Glslang, and SPIRV-Tools
+maintainers work together to handle new HLSL code patterns. This
+document represents the state of the DXC compiler in October 2018.
+
+Glslang does legalization as well. However, what it is able to legalize
+is different from DXC because of features it chooses to support, and the
+optimizations from SPIRV-Tools it choose to run. For example, Glslang
+does not support structured buffer aliasing yet, so many of these
+examples will not work with Glslang.
+
+All of the examples are available in the DXC repository, at
+https://github.com/Microsoft/DirectXShaderCompiler/tree/master/tools/clang/test/CodeGenSPIRV/legal-examples
+. To open a link to Tim Jones' Shader Playground for an example, you can
+follow the url in the comments of each example.
+
+Examples for structured buffers
+===============================
+
+Desired code
+------------
+
+::
+
+    // 0-copy-sbuf-ok.hlsl]
+    // http://shader-playground.timjones.io/e6af2bdce0c61ed07d3a826aa8a95d45
+
+    struct S {
+      float4 f;
+    };
+
+    int i;
+
+    StructuredBuffer<S> gSBuffer;
+    RWStructuredBuffer<S> gRWSBuffer;
+
+    void main() {
+      gRWSBuffer[i] = gSBuffer[i];
+    }
+
+This example shows code that directly translates to valid SPIR-V. In
+this case, we have two structured buffers. When one of their elements is
+accessed, it is done by naming the resource from which to get the
+element.
+
+Note that it is fine to copy an element of the structured buffer.
+
+Single copy to a local
+----------------------
+
+Cases that can be easily legalized are those where there is exactly one
+assignment to the local copy of the structured buffer. In this context,
+a local is either a global static or a function scope symbol. Something
+that can be accessed by only a single instance of the shader. When you
+have a single copy to a local, it is obvious which global is actually be
+used. This allows the compiler to replace a reference to the local
+symbol with the global resource.
+
+Initialization of a static
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+::
```

Comment:
Preferably using `.. code:: hlsl` to highlight the code. See Line 666 at SPIR-V.rst.